### PR TITLE
Discord Bot: Try to use guild emojis for league and race

### PR DESCRIPTION
**Pull request by frugs**
### Summary

This change tries to query the guild emojis to see if the guild has league and race emojis set up (checked by using a simple case-insensitive emoji name match) and tries to use them in the ladder summary if they exist. If they don't, we fall back to the previous behaviour.

This has been implemented as an extra blocking call to try and fetch the guild emojis on every bot command. I have implemented this to use the `EntityRetrievalStrategy.STORE_FALLBACK_REST` strategy which I hope will add some degree of caching to avoid repeated web requests, but I'm not familiar enough with the discord API to have a good understanding of the consequences of this.

### Test plan
I've updated the existing unit test which checks the ladder summary is populated correctly with the stubbed data so that it also creates stubs for guild emojis and checks that these are also correctly rendered. I also downloaded the latest data snapshot into a local db and deployed this locally with the following result:

image

Note that I didn't set up guild emojis for `:master:`, `:protoss:`, and `:zerg:`, and that these are correctly falling back to the previous behaviour.

### CODE REVIEW

**Comment by nephest** [comment]
This is a nice feature to have.

I want to know the reasoning behind this. Is this just for the customization purposes, or the default emojis are bad? I know many ppl don't like the default league badges, but no one complained about races.  Please let me know if this is the case, I'll try to improve the default ones.

## Rebase on 1.45.0
Please rebase your changes on 1.45.0 branch. There should be no conflicts. 1.45.0 requires new aapplication properties to be set: `spring.security.oauth2.client.registration.discord-lg.client-id` and `spring.security.oauth2.client.registration.discord-lg.client-secret`. You should be able to use some stub values if you want.

## Blocking and Caching
`EntityRetrievalStrategy.STORE_FALLBACK_REST` should work as you expect and it's the default strategy. We could use Spring cache since we are performing additional steps like string comparison and map creation. 

### Spring cache
#### Cache creation
The new `getGuildEmojis` method should be moved to [DiscordBootstrap](https://github.com/nephest/sc2-pulse/blob/1.45.0/src/main/java/com/nephest/battlenet/sc2/discord/DiscordBootstrap.java)  class and annotated with Spring's `@Cacheable` annotation. We should probably use guild id as the cache key, so the final annotation should be something like 
 ```java
@Cacheable(cacheNames = "discord-guild-emoji", key="#evt.interaction.guildId.get().asLong()")
```
The `static` modifier must be removed for cache to work.

#### Cache eviction
We should drop the cache when `EmojisUpdateEvent` is received. 

We could add a new empty event handler and annotate it with 
 ```java
@CacheEvict(cacheNames = "discord-guild-emoji", key="#evt.guildId.asLong()")
```
Or just use an empty handler and remove the cache programmatically via the `CacheManager`.

### Additional optimizations
The `HashMap` can be replaced with `EnumMap`, and the method signature can return `Map` or `? extends Map` instead of `HashMap`. 

### Workflow
I can do cache eviction if you are unfamiliar with the Spring framework.

**Comment by frugs** [comment]
> I want to know the reasoning behind this. Is this just for the customization purposes, or the default emojis are bad? I know many ppl don't like the default league badges, but no one complained about races. Please let me know if this is the case, I'll try to improve the default ones.

I'm not sure what you mean by the default ones - do you mean the ones used on the SC2 Pulse website? As this is a discord message, only Unicode emojis and customised Guild Emojis can be shown, and because customised Guild Emojis require a reference to their snowflake ID, we can only use them in messages after querying for them.

To illustrate, this is what I see the SC2 Pulse bot print currently:
image

> rest


I'll try to rebase onto that branch and apply the suggested changes.

**Comment by nephest** [comment]
Yes, I'm talking about the icons from the website, the bot should use them by default. I know that you have to use the emoji id, but the bot should do it by default, it should use the emojis from my default server. Did you give a permission to use external emojis?
 Here is an example
Screenshot 2022-09-07 at 14-41-00 Discord - A New Way to Chat with Friends Communities
Other users from different servers are also able to see the default emojis, or at least that was the case in the past.

**Comment by frugs** [comment]
Interesting - I wasn't aware it was possible to allow this via permissions settings, and thought it was a discord nitro feature.

**Comment by nephest** [comment]
Yes, it requires Nitro for regular user accounts, but bot accounts have access to external emojis for free, probably for this exact use case.

I've made some research. It works if you enable external emojis for `everyone` role. But it doesn't work if the `SC2 Pulse` role have the permission, but `everyone` role does not. Weird, I thought any other role should override the `everyone` role.

**Comment by frugs** [comment]
Thank you, that got it to work! I'll still make your suggested changes and you can decide whether it's worth having this feature or not then :)

**Comment by nephest** [comment]
Sure, I still think it's a nice feature.

**Comment by nephest** [comment]
I found the reason of this behavior https://github.com/discord/discord-api-docs/discussions/3307. tl;dr interactions(slash and user commands in our case) inherit "everyone" role's permissions, but the actual role of the bot is ignored. The issue is marked as resolved on August 8, but seems like the bug persists.

**Comment by frugs** [comment]
Made some significant changes to comply with the suggestions. I think it looks a lot better now!

Most of the heavy lifting has now been moved to a new Component, `GuildEmojiStore`. This now also has its own set of unit tests. It gets autowired into two places - into DiscordBootstrap to facilitate fetching emojis and caching them, and into the SpringDiscordClient to subscribe to `EmojisUpdateEvent` to facilitate cache eviction.

I tested the caching by setting `logging.level.org.springframework.cache=TRACE` and verifying in the logs that the caching works as expected:
```
2022-09-08 02:44:37.192 TRACE 3300 --- [oundedElastic-1] o.s.cache.interceptor.CacheInterceptor   : Computed cache key '219804933916983296' for operation Builder[public java.util.Map com.nephest.battlenet.sc2.discord.GuildEmojiStore.getGuildEmojis(discord4j.core.event.domain.interaction.InteractionCreateEvent)] caches=[discord-guild-emoji] | key='#evt.getInteraction().getGuildId().get()?.asLong()' | keyGenerator='' | cacheManager='' | cacheResolver='' | condition='' | unless='' | sync='false'
2022-09-08 02:44:37.193 TRACE 3300 --- [oundedElastic-1] o.s.cache.interceptor.CacheInterceptor   : Cache entry for key '219804933916983296' found in cache 'discord-guild-emoji'

2022-09-08 02:44:51.173 TRACE 3300 --- [   d4j-events-5] o.s.cache.interceptor.CacheInterceptor   : Computed cache key '219804933916983296' for operation Builder[public reactor.core.publisher.Mono com.nephest.battlenet.sc2.discord.GuildEmojiStore.removeGuildEmojis(discord4j.core.event.domain.guild.EmojisUpdateEvent)] caches=[discord-guild-emoji] | key='#evt.getGuildId().asLong()' | keyGenerator='' | cacheManager='' | cacheResolver='' | condition='',false,false       
2022-09-08 02:44:51.174 TRACE 3300 --- [   d4j-events-5] o.s.cache.interceptor.CacheInterceptor   : Invalidating cache key [219804933916983296] for operation Builder[public reactor.core.publisher.Mono com.nephest.battlenet.sc2.discord.GuildEmojiStore.removeGuildEmojis(discord4j.core.event.domain.guild.EmojisUpdateEvent)] caches=[discord-guild-emoji] | key='#evt.getGuildId().asLong()' | keyGenerator='' | cacheManager='' | cacheResolver='' | condition='',false,false on method public reactor.core.publisher.Mono com.nephest.battlenet.sc2.discord.GuildEmojiStore.removeGuildEmojis(discord4j.core.event.domain.guild.EmojisUpdateEvent)
```

It has been about 10 years since I last worked with Spring, but this seems to work as expected!

File: src/main/java/com/nephest/battlenet/sc2/discord/GuildEmojiStore.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        return Optional.ofNullable(guildEmojis.get(league.getName().toLowerCase(Locale.ENGLISH)));
+    }
+
+    public Optional<String> getGuildRaceEmoji(Map<String, String> guildEmojis, Race race) {
+        return Optional.ofNullable(guildEmojis.get(race.getName().toLowerCase(Locale.ENGLISH)));
+    }
+
+    @Cacheable(cacheNames = "discord-guild-emoji", key="#evt.getInteraction().getGuildId().get()?.asLong()")
+    public Map<String, String> getGuildEmojis(InteractionCreateEvent evt) {
+        return evt.getInteraction()
```
Comment (nephest) [inline src/main/java/com/nephest/battlenet/sc2/discord/GuildEmojiStore.java:44]: I feel like this reactor/stream chain can be improved by splitting the tasks. 
There are 2 tasks here
* Collection of target emoji names
* Collection of emojis themselves

IMO the reactor chain can be more efficient and straightforward if the first task(target emoji names) is moved to a separate unmodifiable static `Set<String>` in [GuildEmojiStore](https://github.com/nephest/sc2-pulse/pull/54/files/3888e1424c9c91154f26bdad5d2c3ab493280de3#diff-4571558ffec0aed6ee411b505b1ec1859b9544f3596411f9d302fc28caa403d7). This way we remove unnecessary map operations(replace and remove), and we are free to use a simple operator chain for emojis like `.flatMapMany().filter(<using static set from the previous step>).collectMap()`.

File: src/test/java/com/nephest/battlenet/sc2/discord/GuildEmojiStoreTest.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    private GuildEmoji terranEmoji;
+
+    @Mock
+    private GuildEmoji otherEmoji;
+
+    private GuildEmojiStore guildEmojiStore;
+
+    @BeforeEach
+    public void beforeEach() {
+        System.out.println("beforeEach");
```
Comment (nephest) [inline src/test/java/com/nephest/battlenet/sc2/discord/GuildEmojiStoreTest.java]: This probably are debugging leftovers and should be removed, or this serves some purpose?

**Review by nephest — CHANGES_REQUESTED** [review]
Looks good! There are 2 minor things that I feel should be changed before I accept this PR(see review comments).

File: src/test/java/com/nephest/battlenet/sc2/discord/GuildEmojiStoreTest.java (hunk 2 above)
Comment (frugs) [reply src/test/java/com/nephest/battlenet/sc2/discord/GuildEmojiStoreTest.java]: This is, as you say, just some debugging leftover when I was mixing test annotations without realising and scratching my head why the before each wasn't being executed. I will remove.

File: src/main/java/com/nephest/battlenet/sc2/discord/GuildEmojiStore.java (hunk 1 above)
Comment (frugs) [reply src/main/java/com/nephest/battlenet/sc2/discord/GuildEmojiStore.java:44]: Ah, very good point! This will help with both efficiency and readability.

**Comment by frugs** [comment]
@nephest I've made the changes you asked for.

**Comment by frugs** [comment]
@nephest I made the changes you suggested, and it's now much clearer what the reactor chain is trying to achieve. Thanks for your help!

**Review by nephest — APPROVED** [review]
Great! Thank you for the contribution!

**Pull request merged**